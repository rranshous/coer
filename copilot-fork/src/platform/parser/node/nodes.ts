/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type { Point, SyntaxNode } from 'web-tree-sitter';
import { BugIndicatingError } from '../../../util/vs/base/common/errors';
import { Range, Uri } from '../../../vscodeTypes';

export interface TreeSitterOffsetRange {
	startIndex: number;
	endIndex: number;
}

export interface TreeSitterExpressionInfo extends TreeSitterOffsetRange {
	version?: number;
	identifier: string;
	text: string;
}

export interface TreeSitterExpressionLocationInfo extends TreeSitterOffsetRange {
	text: string;
	version?: number;
	uri?: Uri;
	range?: Range;
	identifier?: string;
}

export interface TreeSitterPoint {
	row: number;
	column: number;
}

export interface TreeSitterPointRange {
	startPosition: TreeSitterPoint;
	endPosition: TreeSitterPoint;
}

/** Util functions to deal with `TreeSitterOffsetRange` type */
export const TreeSitterOffsetRange = {
	/** check if `container` contains `containee` (non-strict, ie [0, 3] contains [0, 3] */
	doesContain: (container: TreeSitterOffsetRange, containee: TreeSitterOffsetRange): boolean => container.startIndex <= containee.startIndex && containee.endIndex <= container.endIndex,

	ofSyntaxNode: (n: SyntaxNode): TreeSitterOffsetRange => ({ startIndex: n.startIndex, endIndex: n.endIndex }),

	/** sort by `node.startIndex`, break ties by `node.endIndex` (so that nodes with same start index are sorted in descending order) */
	compare: (a: TreeSitterOffsetRange, b: TreeSitterOffsetRange): number => a.startIndex - b.startIndex || b.endIndex - a.endIndex,
	isEqual: (a: TreeSitterOffsetRange, b: TreeSitterOffsetRange): boolean => TreeSitterOffsetRange.compare(a, b) === 0,

	doIntersect: (a: TreeSitterOffsetRange, b: TreeSitterOffsetRange) => {
		const start = Math.max(a.startIndex, b.startIndex);
		const end = Math.min(a.endIndex, b.endIndex);
		return start < end;
	},

	len: (n: TreeSitterOffsetRange) => n.endIndex - n.startIndex,

	/** Given offset ranges [a0, a1] and [b0, b1], returns overlap size */
	intersectionSize: (a: TreeSitterOffsetRange, b: TreeSitterOffsetRange): number => {
		const start = Math.max(a.startIndex, b.startIndex);
		const end = Math.min(a.endIndex, b.endIndex);
		return Math.max(end - start, 0);
	},

	/** Check the given object extends TreeSitterOffsetRange  */
	isTreeSitterOffsetRange(obj: any): obj is TreeSitterOffsetRange {
		return typeof obj.startIndex === 'number' && typeof obj.endIndex === 'number';
	},
};

export const TreeSitterPoint = {

	isEqual(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		return n.row === other.row && n.column === other.column;
	},

	isBefore(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		if (n.row < other.row || (n.row === other.row && n.column < other.column)) {
			return true;
		}
		return false;
	},

	isAfter(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		return TreeSitterPoint.isBefore(other, n);
	},

	isBeforeOrEqual(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		const isBefore = TreeSitterPoint.isBefore(n, other);
		const isEqual = TreeSitterPoint.isEqual(n, other);
		if (isBefore || isEqual) {
			return true;
		}
		return false;
	},

	equals(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		return n.column === other.column && n.row === other.row;
	},

	isAfterOrEqual(n: TreeSitterPoint, other: TreeSitterPoint): boolean {
		return TreeSitterPoint.isBeforeOrEqual(other, n);
	},

	ofPoint: (n: Point): TreeSitterPoint => ({
		row: n.row,
		column: n.column
	}),
};

export const TreeSitterPointRange = {

	/** check if `container` contains `containee` (non-strict) */
	doesContain: (container: TreeSitterPointRange, containee: TreeSitterPointRange): boolean => {
		return TreeSitterPoint.isBeforeOrEqual(container.startPosition, containee.startPosition) && TreeSitterPoint.isAfterOrEqual(container.endPosition, containee.endPosition);
	},

	equals: (a: TreeSitterPointRange, b: TreeSitterPointRange): boolean => {
		return TreeSitterPoint.equals(a.startPosition, b.startPosition) && TreeSitterPoint.equals(a.endPosition, b.endPosition);
	},

	ofSyntaxNode: (n: SyntaxNode): TreeSitterPointRange => ({
		startPosition: n.startPosition,
		endPosition: n.endPosition
	}),
};

export interface Node extends TreeSitterOffsetRange {
	type: string;
}

export const Node = {
	ofSyntaxNode: (n: SyntaxNode) => ({
		type: n.type,
		startIndex: n.startIndex,
		endIndex: n.endIndex,
	}),
};

export interface TreeSitterChunkHeaderInfo extends TreeSitterOffsetRange {
	text: string;
	range: TreeSitterPointRange;
}

export const TreeSitterChunkHeaderInfo = {
	ofSyntaxNode: (n: SyntaxNode): TreeSitterChunkHeaderInfo => ({
		range: TreeSitterPointRange.ofSyntaxNode(n),
		startIndex: n.startIndex,
		text: n.text,
		endIndex: n.endIndex,
	}),
};

/**
 * Represents a node in the overlay tree.
 */
export class OverlayNode {
	constructor(
		public readonly startIndex: number,
		public readonly endIndex: number,
		/**
		 * @example `class_declaration`
		 */
		public kind: string, // TODO@ulugbekna: come up with more generic kinds so that these aren't per-language, then use enum?
		public readonly children: OverlayNode[],
	) {
		if (startIndex > endIndex) {
			throw new BugIndicatingError('startIndex must be less than endIndex');
		}
		let minStartIndex = startIndex;
		for (const child of children) {
			if (child.startIndex < minStartIndex) {
				throw new BugIndicatingError('Invalid child startIndex');
			}
			if (child.endIndex > endIndex) {
				throw new BugIndicatingError('Invalid child endIndex');
			}
			minStartIndex = Math.max(child.endIndex, minStartIndex);
		}
	}

	toString() {
		const printedNodes: string[] = [];
		function toString(node: OverlayNode, indent = '') {
			printedNodes.push(`${indent}${node.kind} [${node.startIndex}, ${node.endIndex}]`);
			node.children.forEach(child => toString(child, indent + '    '));
		}
		toString(this);
		return printedNodes.join('\n');
	}
}
