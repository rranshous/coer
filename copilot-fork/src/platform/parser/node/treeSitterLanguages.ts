/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

/**
 * Languages we can parse using tree-sitter. Each enum member corresponds to a tree-sitter parser.
 */
export enum WASMLanguage {
	Python = 'python',
	JavaScript = 'javascript', // Also includes jsx support
	TypeScript = 'typescript',
	TypeScriptTsx = 'tsx',
	Go = 'go',
	Ruby = 'ruby',
	Csharp = 'csharp',
	Cpp = 'cpp',
	Java = 'java',
	Rust = 'rust',
}

export class TreeSitterUnknownLanguageError extends Error {
	constructor(language: string) {
		super(`Unrecognized language: ${language}`);
	}
}

const languageIdToWasmLanguageMapping: { [language: string]: WASMLanguage } = {
	python: WASMLanguage.Python,
	javascript: WASMLanguage.JavaScript,
	javascriptreact: WASMLanguage.JavaScript,
	jsx: WASMLanguage.JavaScript,
	typescript: WASMLanguage.TypeScript,
	typescriptreact: WASMLanguage.TypeScriptTsx,
	tsx: WASMLanguage.TypeScriptTsx,
	go: WASMLanguage.Go,
	ruby: WASMLanguage.Ruby,
	csharp: WASMLanguage.Csharp,
	cpp: WASMLanguage.Cpp,
	java: WASMLanguage.Java,
	rust: WASMLanguage.Rust,
};

/**
 * @returns a {@link WASMLanguage} if can convert the language ID (from VS Code); otherwise, returns `undefined`.
 */
export function getWasmLanguage(languageId: string): WASMLanguage | undefined {
	if (languageId in languageIdToWasmLanguageMapping) {
		return languageIdToWasmLanguageMapping[languageId];
	}
}
