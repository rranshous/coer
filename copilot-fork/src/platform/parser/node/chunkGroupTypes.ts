/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { TreeSitterChunkHeaderInfo, TreeSitterOffsetRange } from './nodes';

/**
 * For workspace semantic chunking
 */

export interface GenericDetail extends DetailBlock {
	comments: TreeSitterChunkHeaderInfo[];
}
export interface PythonDetail extends DetailBlock {
	docstring?: TreeSitterChunkHeaderInfo;
	decorator?: TreeSitterChunkHeaderInfo;
}

export interface BlockNameDetail extends DetailBlock {
	name: string | undefined;
}

export interface DetailBlock {
	body: TreeSitterChunkHeaderInfo;
}

export interface MatchGroup<DetailType extends DetailBlock> {
	mainBlock: TreeSitterChunkHeaderInfo;
	detailBlocks: DetailType;
}

export interface QueryMatchNode<DetailType extends DetailBlock> {
	info: MatchGroup<DetailType>;
	children: QueryMatchNode<DetailType>[];
}

/**
 * This tree is like the `Parser.SyntaxTree` that we have when we use TreeSitter to parse,
 * but it only has the `MatchGroups` that get passed into the constructor.
 *
 * It sorts the `MatchGroups` into its own hierarchy.
 *
 * This assumes that the constructor's `blockInfos` are a part of the same syntax tree.
 */
export class QueryMatchTree<DetailType extends DetailBlock> {

	public roots: QueryMatchNode<DetailType>[] = [];

	/**
	 * @remark mutates the passed `groups`
	 */
	constructor(groups: MatchGroup<DetailType>[], public readonly syntaxTreeRoot: TreeSitterChunkHeaderInfo) {
		this.formTree(groups);
	}
	/**
	 * This assumes that overlapping blocks imply that one fully overlaps another.
	 * Runs with the same assumptions as `removeOverlapping`.
	 * @param groups to use as node content in the tree
	 */
	private formTree(groups: MatchGroup<DetailType>[]) {
		groups
			.sort((a, b) => a.mainBlock.startIndex - b.mainBlock.startIndex || a.mainBlock.endIndex - b.mainBlock.endIndex);

		const recentParentStack: QueryMatchNode<DetailType>[] = [];

		const peekParent = () => {
			return recentParentStack[recentParentStack.length - 1];
		};

		const hasEqualRange = (a: MatchGroup<DetailType>, b: MatchGroup<DetailType>) => {
			return (a.mainBlock.startIndex === b.mainBlock.startIndex &&
				a.mainBlock.endIndex === b.mainBlock.endIndex);
		};

		for (const group of groups) {
			const matchNode: QueryMatchNode<DetailType> = {
				info: group,
				children: []
			};
			let currParent = peekParent();

			if (!currParent) {
				this.roots.push(matchNode);
				recentParentStack.push(matchNode);
				continue;
			}

			if (hasEqualRange(currParent.info, group)) {
				// any duplicate nodes will always be one after another
				continue;
			}

			while (currParent && !TreeSitterOffsetRange.doesContain(currParent.info.mainBlock, group.mainBlock)) {
				recentParentStack.pop();
				currParent = peekParent();
			}



			if (currParent) {
				currParent.children.push(matchNode);
			} else {
				this.roots.push(matchNode);
			}

			recentParentStack.push(matchNode);
		}
	}

}

/**
 * A tree of block names that would appear in semantic chunks.
 */

export interface BlockNameNode {
	name: string;
	children: BlockNameNode[];
}