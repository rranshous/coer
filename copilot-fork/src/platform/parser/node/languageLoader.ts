/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as path from '../../../util/vs/base/common/path';
import { WASMLanguage } from './treeSitterLanguages';
import Parser = require('web-tree-sitter');

export class LanguageLoader {

	private readonly loadedLanguagesCache;

	constructor() {
		this.loadedLanguagesCache = new Map<WASMLanguage, Promise<Parser.Language>>();
	}

	loadLanguage(wasmLanguage: WASMLanguage): Promise<Parser.Language> {
		if (!this.loadedLanguagesCache.has(wasmLanguage)) {
			this.loadedLanguagesCache.set(wasmLanguage, this._doLoadLanguage(wasmLanguage));
		}
		return this.loadedLanguagesCache.get(wasmLanguage)!;
	}

	private _doLoadLanguage(language: WASMLanguage): Promise<Parser.Language> {
		// construct a path that works both for the TypeScript source, which lives under `/src`, and for
		// the transpiled JavaScript, which lives under `/dist`
		const wasmFileLang = language === 'csharp' ? 'c-sharp' : language;

		const wasmFilename = `tree-sitter-${wasmFileLang}.wasm`;

		// depending on if file is being run from the webpack bundle or source, change the relative path
		const wasmFile =
			path.basename(__dirname) === 'dist'
				? path.resolve(__dirname, wasmFilename)
				: path.resolve(__dirname, '../../../../dist', wasmFilename);

		return Parser.Language.load(wasmFile);
	}
}
