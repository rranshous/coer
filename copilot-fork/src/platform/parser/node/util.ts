/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type { SyntaxNode } from 'web-tree-sitter';
import { WASMLanguage } from './treeSitterLanguages';

/**
 * Extracts the identifier for the given node (is based on heuristics, so not 100% accurate for all languages)
 */
export function extractIdentifier(node: SyntaxNode, languageId: string): string | undefined {
	switch (languageId) {
		case 'python':
		case 'csharp':
			return node.children.find(c => c.type.match(/identifier/))?.text;
		case 'go': {
			const identifierChild = node.children.find(c => c.type.match(/identifier/));
			if (identifierChild) { return identifierChild.text; }
			const specChild = node.children.find(c => c.type.match(/spec/));
			return specChild?.children.find(c => c.type.match(/identifier/))?.text;
		}
		case 'javascript':
		case 'javascriptreact':
		case 'typescript':
		case 'typescriptreact':
		case 'cpp': {
			const declarator = node.children.find(c => c.type.match(/declarator/));
			if (declarator) { return declarator.children.find(c => c.type.match(/identifier/))?.text; }
			const identifierChild = node.children.find(c => c.type.match(/identifier/));
			return identifierChild?.text;
		}
		case 'java': {
			/*
				handles
				```java
				// class identifier
				class Fo<<>>o { }

				class Foo {
					// method identifier
					public void ba<<>>r() { }
				}

				// enum identifier
				enum F<<>>oo {
					// enum constant identifier
					BA<<>>Z
				}

				// interface identifier
				interface Fo<<>>o { }
				```
			*/
			const identifierChild = node.children.find(c => c.type === 'identifier');
			return identifierChild?.text;
		}
		case 'ruby':
			return node.children.find(c => c.type.match(/constant|identifier/))?.text;
		default:
			return node.children.find(c => c.type.match(/identifier/))?.text;
	}
}


export function isDocumentableNode(node: SyntaxNode, language: WASMLanguage) {
	switch (language) {
		case WASMLanguage.TypeScript:
		case WASMLanguage.TypeScriptTsx:
		case WASMLanguage.JavaScript:
			return node.type.match(/definition|declaration|declarator|export_statement/);
		case WASMLanguage.Go:
			return node.type.match(/definition|declaration|declarator|var_spec/);
		case WASMLanguage.Cpp:
			return node.type.match(/definition|declaration|class_specifier/);
		case WASMLanguage.Ruby:
			return node.type.match(/module|class|method|assignment/);
		default:
			return node.type.match(/definition|declaration|declarator/);
	}
}
