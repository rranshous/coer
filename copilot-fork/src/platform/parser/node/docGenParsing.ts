/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type { SyntaxNode } from 'web-tree-sitter';
import { Node, TreeSitterOffsetRange } from './nodes';
import { _parse } from './parserWithCaching';
import { _getNodeMatchingSelection } from './selectionParsing';
import { WASMLanguage } from './treeSitterLanguages';
import { extractIdentifier, isDocumentableNode } from './util';


export type NodeToDocumentContext = {

	/** is undefined when we couldn't determine the identifier */
	nodeIdentifier: string | undefined;

	nodeToDocument: Node;

	/**
	 * 'expanding' - selection was expanded to a wrapping node
	 * 'matchingSelection' - node was picked by observing range overlap with the selection
	 */
	nodeSelectionBy: 'expanding' | 'matchingSelection';
};

/**
 * Starting from the smallest AST node that wraps `selection` and climbs up the AST until it sees a "documentable" node.
 * See {@link isDocumentableNode} for definition of a "documentable" node.
 *
 * @param language The language ID of the source code.
 * @param source The source code to parse.
 * @param selection The range to document.
 * @returns An object containing the smallest node containing the selection range, its parent node, the node to document, and the number of nodes climbed up to reach the documentable node.
 * 			Returns undefined if the language ID is not supported.
 */
export async function _getNodeToDocument(
	language: WASMLanguage,
	source: string,
	selection: TreeSitterOffsetRange
): Promise<NodeToDocumentContext> {

	const treeRef = await _parse(language, source);

	try {

		// if selection is non-empty, try identify a documentable AST node that most matches the selection
		// otherwise, try to find the smallest documentable AST node that wraps the selection

		const isSelectionEmpty = selection.startIndex === selection.endIndex;

		const selectionMatchedNode = isSelectionEmpty ? undefined : _getNodeMatchingSelection(treeRef.tree, selection, language);

		if (selectionMatchedNode) {
			const nodeIdentifier = extractIdentifier(selectionMatchedNode, language);
			return {
				nodeIdentifier,
				nodeToDocument: Node.ofSyntaxNode(selectionMatchedNode),
				nodeSelectionBy: 'matchingSelection'
			};
		}

		const nodeContainingCursor = treeRef.tree.rootNode.descendantForIndex(selection.startIndex, selection.endIndex);

		let nodeToDocument: SyntaxNode = nodeContainingCursor;
		let nNodesClimbedUp = 0;

		// ascend the parse tree until we find a declaration/definition (documentable) node or reach the root node
		while (!isDocumentableNode(nodeToDocument, language) && nodeToDocument.parent !== null) {
			nodeToDocument = nodeToDocument.parent;
			++nNodesClimbedUp;
		}

		const nodeIdentifier = extractIdentifier(nodeToDocument, language);
		return {
			nodeIdentifier,
			nodeToDocument: Node.ofSyntaxNode(nodeToDocument),
			nodeSelectionBy: 'expanding',
		};
	} finally {
		treeRef.dispose();
	}
}

export async function _getDocumentableNodeIfOnIdentifier(
	language: WASMLanguage,
	source: string,
	range: TreeSitterOffsetRange
): Promise<{ identifier: string; nodeRange?: TreeSitterOffsetRange } | undefined> {
	const treeRef = await _parse(language, source);
	try {

		const smallestNodeContainingRange = treeRef.tree.rootNode.descendantForIndex(range.startIndex, range.endIndex);

		if (smallestNodeContainingRange.type.match(/identifier/) &&
			(smallestNodeContainingRange.parent === null || isDocumentableNode(smallestNodeContainingRange.parent, language))
		) {
			const parent = smallestNodeContainingRange.parent;

			const parentNodeRange = parent === null
				? undefined
				: { startIndex: parent.startIndex, endIndex: parent.endIndex };

			return {
				identifier: smallestNodeContainingRange.text,
				nodeRange: parentNodeRange
			};
		}
	} finally {
		treeRef.dispose();
	}
}
