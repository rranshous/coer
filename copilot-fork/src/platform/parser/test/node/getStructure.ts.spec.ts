/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { outdent } from 'outdent';
import { afterAll, describe, expect, it } from 'vitest';
import { resolve } from '../../../../util/vs/base/common/path';
import { _dispose } from '../../node/parserImpl';
import { WASMLanguage } from '../../node/treeSitterLanguages';
import { fromFixture, srcWithAnnotatedStructure } from './getStructure.util';

describe('getStructure - typescript', () => {

	afterAll(() => _dispose());

	function tsSrcWithStructure(source: string) {
		return srcWithAnnotatedStructure(WASMLanguage.TypeScript, source);
	}

	it('`export` keyword should not be visible', async () => {
		const source = outdent`
				export function add(a, b) {
					return a + b;
				}

				function subtract(a, b) {
					return a - b;
				}
				`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<FUNCTION_DECLARATION>export function add(a, b) {
			<RETURN_STATEMENT>	return a + b;
			</RETURN_STATEMENT>}
			</FUNCTION_DECLARATION><FUNCTION_DECLARATION-1>
			function subtract(a, b) {
			<RETURN_STATEMENT-1>	return a - b;
			</RETURN_STATEMENT-1>}</FUNCTION_DECLARATION-1>"
		`);
	});

	it('declarations with comments', async () => {
		const source = outdent`
				class C {
					foo: string; // my comment
				}`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<CLASS_DECLARATION>class C {
			<PUBLIC_FIELD_DEFINITION>	foo: string; // my comment
			</PUBLIC_FIELD_DEFINITION>}</CLASS_DECLARATION>"
		`);
	});

	it('handling line comments', async () => {
		const source = outdent`
			// Example usage
			console.log(farewell("Alice")); // Output: Hello, Alice!
			console.log(farewell("Bob", "Hi")); // Output: Hi, Bob!
			console.log(multiply(5, 1, 3)); // Output: 15
			// Updated identity function to accept only string type
			console.log(identityString("Hello")); // Output: Hello
			// Removed the previous identity function call for number type
			// console.log(identity<number>(42)); // Output: 42
			console.log(identityString("TypeScript")); // Output: TypeScript
			`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<COMMENT>// Example usage
			</COMMENT><EXPRESSION_STATEMENT>console.log(farewell("Alice")); // Output: Hello, Alice!
			</EXPRESSION_STATEMENT><EXPRESSION_STATEMENT-1>console.log(farewell("Bob", "Hi")); // Output: Hi, Bob!
			</EXPRESSION_STATEMENT-1><EXPRESSION_STATEMENT-2>console.log(multiply(5, 1, 3)); // Output: 15
			</EXPRESSION_STATEMENT-2><COMMENT-1>// Updated identity function to accept only string type
			</COMMENT-1><EXPRESSION_STATEMENT-3>console.log(identityString("Hello")); // Output: Hello
			</EXPRESSION_STATEMENT-3><COMMENT-2>// Removed the previous identity function call for number type
			</COMMENT-2><COMMENT-3>// console.log(identity<number>(42)); // Output: 42
			</COMMENT-3><EXPRESSION_STATEMENT-4>console.log(identityString("TypeScript")); // Output: TypeScript</EXPRESSION_STATEMENT-4>"
		`);
	});


	describe('if-else statements', () => {

		it(`capture within statement blocks`, async () => {
			const source = outdent`
					if (true) {
						foo;
					} else {
						bar;
					}
					`;
			expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
				"<IF_STATEMENT>if (true) {
				<EXPRESSION_STATEMENT>	foo;
				</EXPRESSION_STATEMENT>} else {
				<EXPRESSION_STATEMENT-1>	bar;
				</EXPRESSION_STATEMENT-1>}</IF_STATEMENT>"
			`);
		});

		it(`don't capture consequence in: if (cond) statement`, async () => {
			const source = outdent`
					if (true)
						foo;
					else
						bar;
					`;
			expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
				"<IF_STATEMENT>if (true)
					foo;
				else
					bar;</IF_STATEMENT>"
			`);
		});
	});

	it('class declarations', async () => {
		const source = outdent`
				abstract class Animal {
					abstract makeSound(): void;
					move(): void {
						console.log("roaming the earth...");
					}
				}
				class Bear extends Animal {
					constructor(){
						super();
					}
					override makeSound(): void {
						console.log("roar");
					}
				}
				`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<ABSTRACT_CLASS_DECLARATION>abstract class Animal {
			<ABSTRACT_METHOD_SIGNATURE>	abstract makeSound(): void;
			</ABSTRACT_METHOD_SIGNATURE><METHOD_DEFINITION>	move(): void {
			<EXPRESSION_STATEMENT>		console.log("roaming the earth...");
			</EXPRESSION_STATEMENT>	}
			</METHOD_DEFINITION>}
			</ABSTRACT_CLASS_DECLARATION><CLASS_DECLARATION>class Bear extends Animal {
			<CONSTRUCTOR>	constructor(){
			<EXPRESSION_STATEMENT-1>		super();
			</EXPRESSION_STATEMENT-1>	}
			</CONSTRUCTOR><METHOD_DEFINITION-1>	override makeSound(): void {
			<EXPRESSION_STATEMENT-2>		console.log("roar");
			</EXPRESSION_STATEMENT-2>	}
			</METHOD_DEFINITION-1>}</CLASS_DECLARATION>"
		`);
	});

	it('from tree-sitter repo', async () => {
		const source = outdent`
			declare module Foo {
				break;
				continue;
				debugger;
				do { } while (true);
				for (x in null) { }
				for (;;) { }
				if (true) { } else { }
				1;
				return;
				switch (x) {
					case 1:
						break;
					default:
						break;
				}
				throw "hello";
				try { }
				catch (e) { }
				finally { }
			}
			`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<MODULE>declare module Foo {
			<BREAK_STATEMENT>	break;
			</BREAK_STATEMENT><CONTINUE_STATEMENT>	continue;
			</CONTINUE_STATEMENT><DEBUGGER_STATEMENT>	debugger;
			</DEBUGGER_STATEMENT><DO_STATEMENT>	do { } while (true);
			</DO_STATEMENT><FOR_IN_STATEMENT>	for (x in null) { }
			</FOR_IN_STATEMENT><FOR_STATEMENT>	for (;;) { }
			</FOR_STATEMENT><IF_STATEMENT>	if (true) { } else { }
			</IF_STATEMENT><EXPRESSION_STATEMENT>	1;
			</EXPRESSION_STATEMENT><RETURN_STATEMENT>	return;
			</RETURN_STATEMENT><SWITCH_STATEMENT>	switch (x) {
			<SWITCH_CASE>		case 1:
			<BREAK_STATEMENT-1>			break;</BREAK_STATEMENT-1>
			</SWITCH_CASE>		default:
			<BREAK_STATEMENT-2>			break;</BREAK_STATEMENT-2>
				}
			</SWITCH_STATEMENT><THROW_STATEMENT>	throw "hello";
			</THROW_STATEMENT><TRY_STATEMENT>	try { }
				catch (e) { }
				finally { }
			</TRY_STATEMENT>}</MODULE>"
		`);
	});

	it('issue #5755: inline edits go outside the selection', async () => {
		const source = await fromFixture('vscode.proposed.chatParticipantAdditions.d.ts');
		const fileSnapshot = resolve(__dirname, 'fixtures', `vscode.proposed.chatParticipantAdditions-annotated.d.ts.txt`);
		await expect(await tsSrcWithStructure(source)).toMatchFileSnapshot(fileSnapshot);
	});

	it('issue #5755: inline interfaces', async () => {
		const source = outdent`
			interface X {
				doIt(): { x: boolean };
				y: { make(): void };
			}
			`;
		expect(await tsSrcWithStructure(source)).toMatchInlineSnapshot(`
			"<INTERFACE_DECLARATION>interface X {
			<METHOD_SIGNATURE>	doIt(): { x: boolean };
			</METHOD_SIGNATURE><PROPERTY_SIGNATURE>	y: { make(): void };
			</PROPERTY_SIGNATURE>}</INTERFACE_DECLARATION>"
		`);
	});

	describe('handling braces correctly', () => {
		it('if-else chain', async () => {
			const source = outdent`
	function getFirstBracketBefore(node: AstNode, nodeOffsetStart: Length, nodeOffsetEnd: Length, position: Length): IFoundBracket | null {
		if (node.kind === AstNodeKind.List || node.kind === AstNodeKind.Pair) {
			const lengths: { nodeOffsetStart: Length; nodeOffsetEnd: Length }[] = [];
			for (const child of node.children) {
				nodeOffsetEnd = lengthAdd(nodeOffsetStart, child.length);
				lengths.push({ nodeOffsetStart, nodeOffsetEnd });
				nodeOffsetStart = nodeOffsetEnd;
			}
			for (let i = lengths.length - 1; i >= 0; i--) {
				const { nodeOffsetStart, nodeOffsetEnd } = lengths[i];
				if (lengthLessThan(nodeOffsetStart, position)) {
					const result = getFirstBracketBefore(node.children[i], nodeOffsetStart, nodeOffsetEnd, position);
					if (result) {
						return result;
					}
				}
			}
			return null;
		} else if (node.kind === AstNodeKind.UnexpectedClosingBracket) {
			return null;
		} else if (node.kind === AstNodeKind.Bracket) {
			const range = lengthsToRange(nodeOffsetStart, nodeOffsetEnd);
			return {
				bracketInfo: node.bracketInfo,
				range
			};
		}
		return null;
	}
			`;
			expect(await tsSrcWithStructure(source)).toMatchSnapshot();
		});
	});
});
