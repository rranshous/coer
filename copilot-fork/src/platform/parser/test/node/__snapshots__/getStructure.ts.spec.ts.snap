// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getStructure - typescript > handling braces correctly > if-else chain 1`] = `
"<FUNCTION_DECLARATION>function getFirstBracketBefore(node: AstNode, nodeOffsetStart: Length, nodeOffsetEnd: Length, position: Length): IFoundBracket | null {
<IF_STATEMENT>	if (node.kind === AstNodeKind.List || node.kind === AstNodeKind.Pair) {
<LEXICAL_DECLARATION>		const lengths: { nodeOffsetStart: Length; nodeOffsetEnd: Length }[] = [];
</LEXICAL_DECLARATION><FOR_IN_STATEMENT>		for (const child of node.children) {
<EXPRESSION_STATEMENT>			nodeOffsetEnd = lengthAdd(nodeOffsetStart, child.length);
</EXPRESSION_STATEMENT><EXPRESSION_STATEMENT-1>			lengths.push({ nodeOffsetStart, nodeOffsetEnd });
</EXPRESSION_STATEMENT-1><EXPRESSION_STATEMENT-2>			nodeOffsetStart = nodeOffsetEnd;
</EXPRESSION_STATEMENT-2>		}
</FOR_IN_STATEMENT><FOR_STATEMENT>		for (let i = lengths.length - 1; i >= 0; i--) {
<LEXICAL_DECLARATION-1>			const { nodeOffsetStart, nodeOffsetEnd } = lengths[i];
</LEXICAL_DECLARATION-1><IF_STATEMENT-1>			if (lengthLessThan(nodeOffsetStart, position)) {
<LEXICAL_DECLARATION-2>				const result = getFirstBracketBefore(node.children[i], nodeOffsetStart, nodeOffsetEnd, position);
</LEXICAL_DECLARATION-2><IF_STATEMENT-2>				if (result) {
<RETURN_STATEMENT>					return result;
</RETURN_STATEMENT>				}
</IF_STATEMENT-2>			}
</IF_STATEMENT-1>		}
</FOR_STATEMENT><RETURN_STATEMENT-1>		return null;
</RETURN_STATEMENT-1>	} else if (node.kind === AstNodeKind.UnexpectedClosingBracket) {
<RETURN_STATEMENT-2>		return null;
</RETURN_STATEMENT-2>	} else if (node.kind === AstNodeKind.Bracket) {
<LEXICAL_DECLARATION-3>		const range = lengthsToRange(nodeOffsetStart, nodeOffsetEnd);
</LEXICAL_DECLARATION-3><RETURN_STATEMENT-3>		return {
			bracketInfo: node.bracketInfo,
			range
		};
</RETURN_STATEMENT-3>	}
</IF_STATEMENT><RETURN_STATEMENT-4>	return null;
</RETURN_STATEMENT-4>}</FUNCTION_DECLARATION>"
`;
